ssl/quic/quic_txp.c

```diff
@@ -531,115 +606,216 @@ void ossl_quic_tx_packetiser_schedule_ack(OSSL_QUIC_TX_PACKETISER *txp,
+     *       packet, we will discover we have no room to fit it! This is not a
+     *       problem in itself as another datagram can be sent subsequently, but
+     *       it is a problem because we were counting to use that packet to hold
+     *       the essential padding. But if we have already finished encrypting
+     *       the Initial packet, we cannot go and add padding to it anymore.
+     *       This leaves us stuck.
+     *
+     * Because of this, we have to plan multiple packets simultaneously, such
+     * that we can start generating a Handshake (or 0-RTT or 1-RTT, or so on)
+     * packet while still having the option to go back and add padding to the
+     * Initial packet if it turns out to be needed.
+     *
+     * Trying to predict ahead of time (e.g. during Initial packet generation)
+     * whether we will successfully generate a subsequent packet is fraught with
+     * error as it relies on a large number of variables:
+     *
+     *   - Do we have room to fit a packet header? (Consider that due to
+     *     variable-length integer encoding this is highly variable and can even
+     *     depend on payload length due to a variable-length Length field.)
+     *
+     *   - Can we fit even a single one of the frames we want to put in this
+     *     packet in the packet? (Each frame type has a bespoke encoding. While
+     *     our encodings of some frame types are adaptive based on the available
+     *     room - e.g. STREAM frames - ultimately all frame types have some
+     *     absolute minimum number of bytes to be successfully encoded. For
+     *     example, if after an Initial packet there is enough room to encode
+     *     only one byte of frame data, it is quite likely we can't send any of
+     *     the frames we wanted to send.) While this is not strictly a problem
+     *     because we could just fill the packet with padding frames, this is a
+     *     pointless packet and is wasteful.
+     *
+     * Thus we adopt a multi-phase architecture:
+     *
+     *   1. Archetype Selection: Determine desired packet archetype.
+     *
+     *   2. Packet Staging: Generation of packet information and packet payload
+     *      data (frame data) into staging areas.
+     *
+     *   3. Packet Adjustment: Adjustment of staged packets, adding padding to
+     *      the staged packets if needed.
+     *
+     *   4. Commit: The packets are sent to the QTX and recorded as having been
+     *      sent to the FIFM.
+     *
      */
-    cc_can_send = (txp->args.cc_method->get_tx_allowance(txp->args.cc_data) > 0);
+    int res = TX_PACKETISER_RES_FAILURE, rc;
+    uint32_t archetype, enc_level;
+    uint32_t conn_close_enc_level = QUIC_ENC_LEVEL_NUM;
+    struct txp_pkt pkt[QUIC_ENC_LEVEL_NUM];
+    size_t pkts_done = 0;
+    uint64_t cc_limit = txp->args.cc_method->get_tx_allowance(txp->args.cc_data);
 
     for (enc_level = QUIC_ENC_LEVEL_INITIAL;
```

Comment:
As a (future) optimization it would probably make sense in future to record the minimum enc level at which we actually can send packets because all the lower levels have been discarded already and start with that one.